src/main/java/org/javacord/bot/listeners/CommandCleanupListener.java

```diff
@@ -0,0 +1,35 @@
+package org.javacord.bot.listeners;
+
+import org.javacord.api.entity.message.Message;
+import org.javacord.api.entity.message.embed.EmbedBuilder;
+import org.javacord.api.entity.message.embed.EmbedFooter;
+import org.javacord.api.event.message.MessageDeleteEvent;
+import org.javacord.api.listener.message.MessageDeleteListener;
+
+import java.util.Optional;
+
+public class CommandCleanupListener implements MessageDeleteListener {
+
+    public static EmbedBuilder insertResponseTracker(EmbedBuilder builder, long trackedMessageId) {
+        return builder.setFooter("re: " + Long.toString(trackedMessageId));
+    }
+
+    @Override
+    public void onMessageDelete(MessageDeleteEvent event) {
+        Optional<Message> deleted = event.getMessage();
+        if (!deleted.isPresent()) {
+            return;
+        }
+        final String tracker = "re: " + deleted.get().getIdAsString();
+        event.getApi().getThreadPool().getExecutorService().submit(() ->
+                event.getChannel().getMessagesAfterAsStream(deleted.get().getId())
+                        .filter(msg -> msg.getEmbeds().size() == 1)
```

Comment:
JDK9 `Stream#takeWhile` would be perfect for this. I haven't found a suitable replacement so for now I'll just limit it to 100 messages.

Reply:
At least from the JavaDoc it seems I order them from first given id to second given id, so that should not be the cause.
If you think the problem is, that the calculated message id does not exist, this shouldn't be a problem at all, it should work nevertheless.

Even if using the `takeWhile`, you shouldn't calculate the creation timestamp for each message, but again calculate the boundary message ID and just compare the ID.

While I think the `getMessagesBetweenAsStream` approach would be better as we right away only request the messages from Discord that we might be interested in and Discord does the filtering, you could also do a Java-9-similar approach by using `getMessagesAfterWhileAsStream` or `getMessagesAfterUntilAsStream` and again comparing the id against the calculated boundary id.

And having said all this, we can also consider running the bot on Java 11 if you like, I have absolutely no problem with that.

Reply:
I do feel like the `getMessagesBetweenAsStream` approach is hacky and programming a bot should not depend on that level of understanding of the ID schema. The `getMessagesAfterUntil` approach I do like much more, assuming the API correctly stops requesting after the requirements are met.

I have not yet tested this new approach but pushed it already for your consideration. It does, however, feel infinitly more clean than the old one and does not leave the bitter aftertaste that seemingly arbitrarily constructing a boundary ID would have.

Reply:
> I do feel like the getMessagesBetweenAsStream approach is hacky and programming a bot should not depend on that level of understanding of the ID schema. The getMessagesAfterUntil approach I do like much more, assuming the API correctly stops requesting after the requirements are met.

Well, I wouldn't call it hacky, but a performance optimization by insight.
But you are probably right, it is better to code nicer unless you actually have a performance issue to solve and I doubt we have, even with calculating the timestamp from each ID, as usually the answer should come fast after the triggering message and you abort when found.

The API should stop querying if the condition is met indeed how you wrote it. The code done internally is `getMessagesAsStream(channel, before, after).peek(messages::add).filter(condition).findFirst();`, so it stops processing the stream as soon as one message is found that mets the condition and the stream retrieves the messages in batches of 100 as needed, so I think your solution should work fine.

> I have not yet tested this new approach but pushed it already for your consideration. It does, however, feel infinitly more clean than the old one and does not leave the bitter aftertaste that seemingly arbitrarily constructing a boundary ID would have.

I didn't test it either, but it looks good to me. Maybe the small nit-pick I added to only calculate the tracker once and have a method reference then in the `filter`.